src/sql/workbench/contrib/notebook/find/notebookFindModel.ts

```diff
@@ -523,43 +523,64 @@ export class NotebookFindModel extends Disposable implements INotebookFindModel
 		return this.findArray;
 	}
 
-	private searchFn(cell: ICellModel, exp: string, maxMatches?: number): NotebookRange[] {
+	private searchFn(cell: ICellModel, exp: string, matchCase: boolean = false, wholeWord: boolean = false, maxMatches?: number): NotebookRange[] {
 		let findResults: NotebookRange[] = [];
 		let cellVal = cell.cellType === 'markdown' ? this.cleanUpCellSource(cell.source) : cell.source;
-		let index: number;
-		let start: number;
-		let end: number;
 		if (cellVal) {
+
 			if (typeof cellVal === 'string') {
-				index = 0;
-				while (cellVal.substr(index).toLocaleLowerCase().indexOf(exp.toLocaleLowerCase()) > -1) {
-					start = cellVal.substr(index).toLocaleLowerCase().indexOf(exp.toLocaleLowerCase()) + index;
-					end = start + exp.length;
-					let range = new NotebookRange(cell, 0, start, 0, end);
-					findResults = findResults.concat(range);
-					index = end;
-				}
+				let findStartResults = this.search(cellVal, exp, matchCase, wholeWord, maxMatches);
+				findStartResults.forEach(start => {
+					let range = new NotebookRange(cell, 0, start, 0, start + exp.length);
+					findResults.push(range);
+				});
+
 			} else {
 				for (let j = 0; j < cellVal.length; j++) {
-					index = 0;
 					let cellValFormatted = cell.cellType === 'markdown' ? this.cleanMarkdownLinks(cellVal[j]) : cellVal[j];
-					while (cellValFormatted.substr(index).toLocaleLowerCase().indexOf(exp.toLocaleLowerCase()) > -1) {
-						start = cellValFormatted.substr(index).toLocaleLowerCase().indexOf(exp.toLocaleLowerCase()) + index + 1;
-						end = start + exp.length;
+					let findStartResults = this.search(cellValFormatted, exp, matchCase, wholeWord, maxMatches - findResults.length);
+					findStartResults.forEach(start => {
 						// lineNumber: j+1 since notebook editors aren't zero indexed.
-						let range = new NotebookRange(cell, j + 1, start, j + 1, end);
-						findResults = findResults.concat(range);
-						index = end;
-					}
-					if (findResults.length >= maxMatches) {
-						break;
-					}
+						let range = new NotebookRange(cell, j + 1, start, j + 1, start + exp.length);
+						findResults.push(range);
+					});
 				}
 			}
 		}
 		return findResults;
 	}
 
+	escapeRegExp(text: string): string {
+		return text.replace(/[-[\]{}()*+!<=:?.\/\\^$|#\s,]/g, '\\$&');
+	}
+
+	search(input: string, exp: string, matchCase: boolean = false, wholeWord: boolean = false, maxMatches?: number): number[] {
+		let index: number = 0;
+		let start: number;
+		let findResults: number[] = [];
+		let wholeWordRegex = new RegExp(`\\b${this.escapeRegExp(exp)}\\s`);
+		if (!matchCase) {
+			input = input.toLocaleLowerCase();
+			exp = exp.toLocaleLowerCase();
+			wholeWordRegex = new RegExp(`\\b${this.escapeRegExp(exp)}\\s`);
```

Comment:
You could create this regex just after the **if** instead, since it only depends on **exp** and doesn't get used in the **if**.

Reply:
Better to leave it here IMO - no reason to recreate the regex if we don't need to (note that this is just overwriting the original regex).

What I would suggest though is having the original definition in an else instead so that we only create one instance no matter what.

Reply:
Wouldn't need an else, since it doesn't get used until after this point. The if case would just set **input** and **exp**, and then the regex could be created after.

Reply:
Yup, good point